src/prototype_creep_harvest.js

```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
   for (const container of containers) {
     resourceAtPosition += _.sum(container.store);
   }
-
-  if (resourceAtPosition > carryCapacity) {
-    Game.rooms[this.memory.base].checkRoleToSpawn('carry', 0, this.memory.routing.targetId, this.memory.routing.targetRoom);
-  } else if (resourceAtPosition <= HARVEST_POWER * workParts) {
+  if (resourceAtPosition > parts.carryParts.carry * CARRY_CAPACITY) {
+    Game.rooms[this.memory.base].checkRoleToSpawn('carry', 0, this.memory.routing.targetId, this.memory.routing.targetRoom, carrySettings);
+  } else if (resourceAtPosition <= HARVEST_POWER * parts.sourcerWork) {
     const nearCarries = this.pos.findInRangePropertyFilter(FIND_MY_CREEPS, 2, 'memory.role', ['carry'], false, {
       filter: (creep) => creep.memory.routing.targetId === this.memory.routing.targetId,
     });
     if (nearCarries.length > 1) {
       nearCarries[0].memory.recycle = true;
     }
   }
-  this.memory.wait = Math.max(waitTime, config.carry.minSpawnRate);
+  this.memory.wait = this.getCarrySpawnInterval(parts, resourceAtPosition);
+};
+
+/*
+Time between carrys should be proportional to % of energy the carry will carry in his life as :
+Energy harvested in sourcer life :
+`1500 * (HARVEST_POWER * workParts) = A`
+Energy carried by carry :
+`1500 * carryCapacity / carryTravelTime = B`
+`B/A = carryCapacity /( carryTravelTime * harvestpower * workParts) `
+This result should be interpreted as spawn a creep each 1500 ticks is B/A = 1. One each 750 ticks if it's 1/2.
+then we just have to multiply it by 1500 and floor all for have a rounded value.
+The distance coeff is for take care of surpopulation in parents rooms, then far rooms will call less carry. Not even sure we need it.
+*/
+
+Creep.prototype.getCarrySpawnInterval = function(parts, resourcesDroped) {
+  if (!this.memory.pathDatas) {
+    this.memory.pathDatas = {plain: 10};
+  }
+
+  const {sourcerWork, carryParts} = parts;
+  let travelTime = 0;
+
+  let terrain;
+  const terrainCost = {plain: 1, swamp: 5, road: 0.5};
+  for (terrain of ['plain', 'swamp', 'road']) {
+    travelTime += this.memory.pathDatas[terrain] * Math.max(1, terrainCost[terrain] * Math.ceil(carryParts.work / carryParts.move));
+    travelTime += this.memory.pathDatas[terrain] * Math.max(1, terrainCost[terrain] * Math.ceil((carryParts.work + carryParts.carry) / carryParts.move));
+  }
+
+  const sourceFullFilled = Game.getObjectById(this.memory.routing.targetId).energyCapacity / 3000;
```

Comment:
If you want to take a source metric into account you need to compare it to the time to regenerate. `energyCapacity = 10` is totally fine if it takes only 1 tick to be filled up again.